src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1075,6 +1080,45 @@ internal OneOrMany<SyntaxTree> GetSyntaxTreesByMappedPath(string mappedPath)
             }
         }
 
+        private class ContentHashComparer : IEqualityComparer<ReadOnlyMemory<byte>>
+        {
+            public bool Equals(ReadOnlyMemory<byte> x, ReadOnlyMemory<byte> y)
+            {
+                return x.Span.SequenceEqual(y.Span);
+            }
+
+            public int GetHashCode(ReadOnlyMemory<byte> obj)
+            {
+                return BinaryPrimitives.ReadInt32LittleEndian(obj.Span);
+            }
+
+            public static ContentHashComparer Instance { get; } = new ContentHashComparer();
+        }
+
+        internal OneOrMany<SyntaxTree> GetSyntaxTreesByContentHash(ReadOnlyMemory<byte> contentHash)
+        {
+            var contentHashToSyntaxTree = _contentHashToSyntaxTree;
+            if (contentHashToSyntaxTree.IsDefault)
+            {
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _contentHashToSyntaxTree, computeHashToSyntaxTree());
+                contentHashToSyntaxTree = _contentHashToSyntaxTree;
+            }
+
+            return contentHashToSyntaxTree.TryGetValue(contentHash, out var value) ? value : OneOrMany<SyntaxTree>.Empty;
+
+            ImmutableSegmentedDictionary<ReadOnlyMemory<byte>, OneOrMany<SyntaxTree>> computeHashToSyntaxTree()
+            {
+                var builder = ImmutableSegmentedDictionary.CreateBuilder<ReadOnlyMemory<byte>, OneOrMany<SyntaxTree>>(ContentHashComparer.Instance);
+                foreach (var tree in SyntaxTrees)
+                {
+                    var text = tree.GetText();
```

Comment:
can cancellation be passed in here? is it something that can be threaded through by the caller?

Reply:
It's called through `DecodeWellKnownAttributeImpl` on symbols, not practical to thread cancellation through.

Reply:
only when someone uses this well known attribute type though, right?  otherwise, you won't pay the price to get all these texts and checksums?

Reply:
Exactly, we have to make it all the way to seeing that they are using InterceptsLocationAttribute before we do this work.